docs/openfaas-pro/jetstream.md

```diff
@@ -18,17 +19,155 @@ On our blog we demo and explore some architectural patterns for these uses cases
+
+For production environments you will need to install NATS separately using its Helm chart with at least 3 server replicas, so that if a pod crashes, the data can be recovered automatically.
+
+```yaml
+queueMode: jetstream
+nats:
+  streamReplication: 3
+  external:
+    enabled: true
+    host: "nats.nats"
+    port: "4222"
+```
+
+## Configure JetStream
+
+Every OpenFaaS async queue requires a Stream and Consumer to be created on the JetStream server. The queue-worker creates these on first startup if they do not exist. The default Stream and Consumer have some limits:
+
+- The maximum concurrency of each queue is limited to 512.
+- There is a limit on the amount of messages that can be queued for retries. The queue-worker will stop processing new messages when this limit is reached.
+
+To extend these limits Streams and Consumers can be created manually.
+
+### Configure Streams and Consumers manually
+
+Streams and Consumers can be defined manually, typically using the [NATS CLI tool](https://docs.nats.io/running-a-nats-service/configuration/resource_management/configuration_mgmt/nats-admin-cli). 
+
+>A [Kubernetes controller](https://docs.nats.io/running-a-nats-service/configuration/resource_management/configuration_mgmt/kubernetes_controller) is available for managing Streams and Consumers declaratively. This can be used if you are using a CD system like ArgoCD or Flux.
+
+You can use [arkade](https://github.com/alexellis/arkade) to install the NATS CLI.
+```
+arkade get nats
+```
+
+Port forward the nats server to your localhost so you can use the cli to interact with it.
+```bash
+kubectl port-forward -n openfaas svc/nats 4222:4222
+```
+#### Create a Stream
+
+The Stream will need to be created first. In this example we will create the Stream for the shared queue, `faas-request`.
+
+```bash
+nats stream create faas-request \
+  --replicas=1 \
+  --subjects=faas-request \
+  --storage=file \
+  --retention=limits \
+  --discard=old \
+  --max-msgs=-1 \
+  --max-msgs-per-subject=-1 \
+  --max-bytes=-1 \
+  --max-age=-1 \
+  --max-msg-size=-1 \
+  --dupe-window="2m" \
+  --no-allow-rollup \
+  --no-deny-delete \
+  --no-deny-purge
+```
+
+Messages intended for a queue are published to a NATS subject, `faas-request` by default and the queue name for a [dedicated queue](/reference/async/#multiple-queues). A Stream should only consume the subject for the queue that it will be associated with. This can be configured with the `--subjects` flag.
+
+The `--replicas` flag is used to configure the stream replication factor. This should be at least 3 for production environments.
+
+#### Create a Consumer
+
+Once the Stream has been created the Consumer can be added. We will create a consumer `faas-workers` for the `faas-request` stream.
+
+```bash
+nats consumer \
+  create faas-request faas-workers \
+  --pull \
+  --ack=explicit \
+  --deliver=all \
+  --replay=instant \
+  --wait=3m \
```

Comment:
I explained the parameters. Users are prompted with the remaining required values if they are not specified. Because we use some of the default settings but in other prompts a value needs to be selected from a list of options.

I think setting the flags explicitly is more clear.

Reply:
> I think setting the flags explicitly is more clear.

It may be clearer to some people, and overwhelming to others.

When we create the stream/consumer the customer has absolutely no concept of any of these flags, then we show them all of these and force them to decide if the numbers make sense.

If there are any that can be omitted, where it doesn't ask for a prompt, please do so.

Reply:
The long list of arguments is required if you don't want any prompts. With this short list there will be prompts but the defaults can be selected.

Reply:
I don't want any prompts if it can be avoided.

Reply:
I changed this to include the minimum required flags to not get any prompts.